app/scripts/modules/core/src/pipeline/config/stages/entityTags/ApplyEntityTagsStageConfig.tsx

```diff
@@ -0,0 +1,135 @@
+import * as React from 'react';
+import Select, { Option } from 'react-select';
+
+import { IEntityRef, IEntityTag, IStage } from 'core/domain';
+import { IStageConfigProps, StageConfigField } from 'core/pipeline';
+import { SETTINGS } from 'core/config/settings';
+
+import { TagEditor } from './TagEditor';
+import { FormField, ReactSelectInput, StringsAsOptions } from 'core/presentation';
+
+export interface IApplyEntityTagsStage extends IStage {
+  entityRef: IEntityRef;
+  tags: IEntityTag[];
+}
+
+export interface IApplyEntityTagsStageConfigProps extends IStageConfigProps {
+  stage: IApplyEntityTagsStage;
+}
+
+export class ApplyEntityTagsStageConfig extends React.Component<IApplyEntityTagsStageConfigProps> {
+  private entityOptions = [
+    { label: 'Server Group', value: 'servergroup' },
+    { label: 'Load Balancer', value: 'loadbalancer' },
+    { label: 'Application', value: 'application' },
+    { label: 'Cluster', value: 'cluster' },
+    { label: 'Security Group', value: 'securitygroup' },
+  ];
+
+  private entityRefTypeChanged = (entityType: string) => {
+    const entityRef = { entityType };
+    this.props.updateStageField({ entityRef });
+  };
+
+  private entityRefFieldChanged = (key: string, value: string) => {
+    const entityRef = { ...this.props.stage.entityRef, ...{ [key]: value } };
+    this.props.updateStageField({ entityRef });
+  };
+
+  private tagChanged = (tag: IEntityTag, index: number) => {
+    const tags = this.props.stage.tags.slice();
+    tags[index] = tag;
+    this.props.updateStageField({ tags });
+  };
+
+  private addTag = () => {
+    this.props.stage.tags.push({ name: '', value: '' });
+    this.props.updateStageField({ tags: this.props.stage.tags });
```

Comment:
To avoid  array mutation?
```suggestion
    this.props.updateStageField({ tags: this.props.stage.tags.concat({ name: '', value: '' }) });
```

Reply:
👍